dag_in_context/src/pretty_print.rs

```diff
@@ -0,0 +1,831 @@
+use crate::{
+    from_egglog::FromEgglog,
+    prologue,
+    schema::{
+        self, Assumption, BaseType, BinaryOp, Expr, RcExpr, TernaryOp, TreeProgram, Type, UnaryOp,
+    },
+    schema_helpers::AssumptionRef,
+    to_egglog::TreeToEgglog,
+};
+use egglog::{Term, TermDag};
+
+use std::{collections::HashMap, hash::Hash, rc::Rc, vec};
+
+pub struct PrettyPrinter {
+    pub expr: RcExpr,
+    // Type/Assum/BaseType -> intermediate variables
+    symbols: indexmap::IndexMap<Symbols, String>,
+    // intermediate variables about to print
+    log: Vec<String>,
```

Comment:
Why do you need a log? I think the `table` should give you the order things should be printed if you use an indexmap

Reply:
mainly bacause Expr cannot be hash so I cannot use a indexmap for table. So need a log to keep track of order or insert.